### ISSUE TITLE
create a service account and setIamPolicy fails

### ISSUE THREAD

**Issue by hilliao** [GoogleCloudPlatform/deploymentmanager-samples#441]
I'm following the examples in [google-deployment-manager-assigning-iam-policies-at-project](https://stackoverflow.com/questions/48710548/google-deployment-manager-assigning-iam-policies-at-project) article. I have the following python and jinja templates:
**Python template** service-accounts.py
```
def GenerateConfig(context):
    project_id = context.env['project']

    resources = [
        {
            'name': 'svc-account-test',
            'type': 'iam.v1.serviceAccount',
            'properties': {
                'accountId': 'svc-account-test',
                'displayName': 'svc-account-test',
                'projectId': project_id
            }
        },
        {
            'name': 'get-iam-policy',
            'action': 'gcp-types/cloudresourcemanager-v1:cloudresourcemanager.projects.getIamPolicy',
            'properties': {
                'resource': project_id,
            },
            'metadata': {
                'runtimePolicy': ['UPDATE_ALWAYS']
            }
        },
        {
            'name': 'patch-iam-policy',
            'action': 'gcp-types/cloudresourcemanager-v1:cloudresourcemanager.projects.setIamPolicy',
            'properties': {
                'resource': project_id,
                'policy': '$(ref.get-iam-policy)',
                'gcpIamPolicyPatch': {
                    'add': [
                        {
                            'role': 'roles/dataflow.admin',
                            'members': [
                                {
                                    'serviceAccount': 'svc-account-test@my_proj_id.iam.gserviceaccount.com'
                                }
                            ]
                        }
                    ]
                }
            }
        }
    ]
```
**jinja templates** service-accounts.jinja
```
    return {'resources': resources}

{% set deployment = env['deployment'] %}
{% set project = env['project'] %}

resources:
- name: {{ deployment }}-svc-account
  type: iam.v1.serviceAccount
  properties:
    accountId: svc-account-test
    displayName: svc-account-test
- name: get-iam-policy
  action: gcp-types/cloudresourcemanager-v1:cloudresourcemanager.projects.getIamPolicy
  properties:
    resource: {{ project }}
  metadata:
    runtimePolicy:
    - 'UPDATE_ALWAYS'
- name: patch-iam-policy
  action: gcp-types/cloudresourcemanager-v1:cloudresourcemanager.projects.setIamPolicy
  properties:
    resource: {{ project }}
    policy: $(ref.get-iam-policy)
    gcpIamPolicyPatch:
      add:
      - role: roles/bigquery.dataEditor
        members:
        - serviceAccount: svc-account-test@my_proj_id.iam.gserviceaccount.com

```
**service-accounts.yaml** where .py can be replaced with .jinja
```
imports:
  - path: ../templates/service-accounts.py

resources:
  - name: service-accounts
    type: ../templates/service-accounts.py
    properties:
      project: my_proj_id
```
I consistently got the following error:
```
ERROR: (gcloud.deployment-manager.deployments.create) Error in Operation [operation-1556754002417-587dc0951192f-ea2abc18-50b7b1a1]: errors:
- code: CONDITION_NOT_MET
  location: /deployments/hil-test-svc12/resources/patch-iam-policy->$.properties->$.policy
  message: |-
    InputMapping for field [policy] for method [setIamPolicy] could not be set from input, mapping was: [$.gcpIamPolicyPatch($.resource.properties.policy, $.resource.properties.gcpIamPolicyPatch)], and evaluation context was:
    {
      "deployment" : {
        "id" : 1175707334797136061,
        "name" : "hil-test-svc12"
      },
      "extensions" : {
        "EnableAdditionalJsonPathFunctions" : true,
        "EnableGoogleTypeProviderFunctionsExperiment" : true
      },
      "intent" : "CREATE",
      "matches" : [ ],
      "project" : "my_proj_id",
      "requestId" : "3b575cb5-9d31-3175-9f5a-576326b4cbc4",
      "resource" : {
        "name" : "patch-iam-policy",
        "properties" : {
          "gcpIamPolicyPatch" : {
            "add" : [ {
              "members" : [ {
                "serviceAccount" : "svc-account-test@my_proj_id.iam.gserviceaccount.com"
              } ],
              "role" : "roles/dataflow.admin"
            } ]
          },
          "policy" : {
            "bindings" : [ {
              "members" : [ "serviceAccount:000111222333@cloudbuild.gserviceaccount.com", "serviceAccount:local-runner@my_proj_id.iam.gserviceaccount.com" ],
              "role" : "roles/bigtable.admin"
            }, {
              "members" : [ "serviceAccount:000111222333@cloudbuild.gserviceaccount.com" ],
              "role" : "roles/cloudbuild.builds.builder"
            }, {
              "members" : [ "serviceAccount:service-000111222333@gcf-admin-robot.iam.gserviceaccount.com" ],
              "role" : "roles/cloudfunctions.serviceAgent"
            }, {
              "members" : [ "serviceAccount:service-000111222333@compute-system.iam.gserviceaccount.com" ],
              "role" : "roles/compute.serviceAgent"
            }, {
              "members" : [ "serviceAccount:my_proj_id-svc-account@my_proj_id.iam.gserviceaccount.com" ],
              "role" : "roles/container.admin"
            }, {
              "members" : [ "serviceAccount:my_proj_id-svc-account@my_proj_id.iam.gserviceaccount.com" ],
              "role" : "roles/container.developer"
            }, {
              "members" : [ "serviceAccount:service-000111222333@container-engine-robot.iam.gserviceaccount.com" ],
              "role" : "roles/container.serviceAgent"
            }, {
              "members" : [ "serviceAccount:000111222333@cloudbuild.gserviceaccount.com", "serviceAccount:dataflow-admin@my_proj_id.iam.gserviceaccount.com", "serviceAccount:local-runner@my_proj_id.iam.gserviceaccount.com" ],
              "role" : "roles/dataflow.admin"
            }, {
              "members" : [ "serviceAccount:service-000111222333@dataflow-service-producer-prod.iam.gserviceaccount.com" ],
              "role" : "roles/dataflow.serviceAgent"
            }, {
              "members" : [ "serviceAccount:000111222333@cloudbuild.gserviceaccount.com", "serviceAccount:local-runner@my_proj_id.iam.gserviceaccount.com" ],
              "role" : "roles/dataflow.worker"
            }, {
              "members" : [ "serviceAccount:000111222333@cloudbuild.gserviceaccount.com" ],
              "role" : "roles/deploymentmanager.editor"
            }, {
              "members" : [ "serviceAccount:000111222333-compute@developer.gserviceaccount.com", "serviceAccount:my_proj_id@appspot.gserviceaccount.com", "serviceAccount:service-000111222333@containerregistry.iam.gserviceaccount.com", "user:mukesh.kumar@mavenwave.com", "user:nikhil.rana@mavenwave.com", "user:priyanka.bhatia@mavenwave.com", "user:puneet.khera@mavenwave.com", "user:sumit.bhakt@mavenwave.com", "user:udayan.tat@mavenwave.com" ],
              "role" : "roles/editor"
            }, {
              "members" : [ "serviceAccount:service-000111222333@firebase-rules.iam.gserviceaccount.com" ],
              "role" : "roles/firebaserules.system"
            }, {
              "members" : [ "serviceAccount:service-000111222333@gcp-sa-pubsub.iam.gserviceaccount.com" ],
              "role" : "roles/iam.serviceAccountTokenCreator"
            }, {
              "members" : [ "serviceAccount:000111222333@cloudservices.gserviceaccount.com", "user:hil.liao@mavenwave.com", "user:jess.lampe@mavenwave.com", "user:manoj.kumar@mavenwave.com", "user:sumeet.singh@mavenwave.com" ],
              "role" : "roles/owner"
            }, {
              "members" : [ "serviceAccount:000111222333@cloudbuild.gserviceaccount.com", "serviceAccount:local-runner@my_proj_id.iam.gserviceaccount.com", "serviceAccount:my_proj_id-svc-account@my_proj_id.iam.gserviceaccount.com" ],
              "role" : "roles/pubsub.publisher"
            }, {
              "members" : [ "serviceAccount:000111222333@cloudbuild.gserviceaccount.com", "serviceAccount:local-runner@my_proj_id.iam.gserviceaccount.com" ],
              "role" : "roles/pubsub.subscriber"
            }, {
              "members" : [ "serviceAccount:000111222333@cloudbuild.gserviceaccount.com", "serviceAccount:local-runner@my_proj_id.iam.gserviceaccount.com" ],
              "role" : "roles/pubsub.viewer"
            }, {
              "members" : [ "serviceAccount:service-000111222333@serverless-robot-prod.iam.gserviceaccount.com" ],
              "role" : "roles/run.serviceAgent"
            }, {
              "members" : [ "serviceAccount:000111222333@cloudbuild.gserviceaccount.com" ],
              "role" : "roles/serviceusage.serviceUsageAdmin"
            }, {
              "members" : [ "serviceAccount:000111222333@cloudbuild.gserviceaccount.com" ],
              "role" : "roles/source.reader"
            }, {
              "members" : [ "serviceAccount:service-000111222333@sourcerepo-service-accounts.iam.gserviceaccount.com" ],
              "role" : "roles/sourcerepo.serviceAgent"
            }, {
              "members" : [ "serviceAccount:my_proj_id-svc-account@my_proj_id.iam.gserviceaccount.com" ],
              "role" : "roles/storage.objectViewer"
            }, {
              "members" : [ "serviceAccount:local-runner@my_proj_id.iam.gserviceaccount.com" ],
              "role" : "roles/viewer"
            } ],
            "etag" : "BwWH224BIPE=",
            "version" : 1
          },
          "resource" : "my_proj_id"
        },
        "self" : { }
      }
    }
```
project ID and number are stripped out.

**Comment by ocsig** [comment]
TLDR: You need to create the SA first, only then can you do the IAM Bindings, you need a reference/dependsON.

Your deployment is failing because you are asking DM to create 3 resources: 1 Service accont and 2 IAM binding, however you do not use any [Deployment Manager refferences](https://cloud.google.com/deployment-manager/docs/step-by-step-guide/using-references), so DM doesn't know, there is a strict execution order between these 3 resources.

I highly recommend for you to check out the CloudFoundation Toolkit folder ( and the new repo where we are corrently transitioning).

[CFT IAM member binding NEW repo](https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/tree/master/dm/templates/iam_member)

[CFT IAM member binding OLD repo](https://github.com/GoogleCloudPlatform/deploymentmanager-samples/tree/master/community/cloud-foundation/templates/iam_member)

[CFT Project creation Factory NEW repo](https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/tree/master/dm/templates/project) (this containts a full SA + IAM binding example.

**Comment by hilliao** [comment]
Thanks for the tip. I used your new type and successfully created a service account and added IAM policy bindings to it. I shared the example at [google-deployment-manager-assigning-iam-policies-at-project article](https://stackoverflow.com/questions/48710548/google-deployment-manager-assigning-iam-policies-at-project/55961393#55961393). If you think that's valuable, please vote my answer up. Thanks!
